cmd/gen-go-sample/main.go

```diff
@@ -203,6 +282,35 @@ func (g *generator) commit(gofmt bool, year int) ([]byte, error) {
 	return b, nil
 }
 
+func (g *generator) disambiguateSampleIDs() error {
+	idCount := make(map[string]int)
+	samples := g.sampleConfig.Samples
+	for i := range samples {
+		// default ID to region tag
+		if samples[i].ID == "" {
+			samples[i].ID = samples[i].RegionTag
+		}
+		if samples[i].ID != "" {
+			idCount[samples[i].ID]++
+		}
+	}
+
+	for i := range samples {
+		// Generate a sample ID when the user-provided ID is empty not unique
+		if samples[i].ID == "" || idCount[samples[i].ID] > 1 {
+			jsonStr, err := json.Marshal(samples[i])
+			if err != nil {
+				return err
+			}
+			checkSum := sha256.Sum256([]byte(jsonStr))
```

Comment:
My reasoning was the same: In that case there are two exactly identical sample configuration, thus just generating one sample isn't a bad outcome.

Reply:
The python microgenerator is susceptible to the same non-uniqueness scenario.
I would argue it's not _wrong_ to do the work twice and overwrite, but it it would definitely be more user friendly to fail loudly.

Reply:
"Failing loudly" in this case would entail keeping the hash as is (without appending the `i`) and then checking that each hash is unique. (Agreed?)

I can see pros and cons. I wonder whether a warning rather than an error would be better: we still generate one sample but we warn the user that two separate configs resulted in the same sample. I can imagine that in some cases (eg during sample configuration, which might involve a lot of copying and pasting and testing) that might be acceptable.

Reply:
Pretty much; by "fail loudly" I mean raise a terminating error of some kind.
Not to get back on this soap box, but I think a warning is a compromise nobody wants. Silently swallowing duplicates and erroring out on duplicates are both consistent, reasonable choices.
Flagging warnings feels like a slippery slope to being too clever, like when you pointed out that calling form wasn't sufficient to disambiguate duplicated IDs. Instead of trying to auto-generate a descriptive, unique ID and falling back to appending a hash, it was a lot easier and less error prone to just disambiguate using hashes immediately.

Reply:
Done.